Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/ORTHelpers.cpp

```diff
@@ -0,0 +1,157 @@
+#include "pch.h"
+#include "ORTHelpers.h"
+#undef min
+
+#define THROW_IF_FAILED(hr) {HRESULT localHr = (hr); if (FAILED(hr)) throw hr;}
+#define RETURN_IF_FAILED(hr) {HRESULT localHr = (hr); if (FAILED(hr)) return hr;}
+#define THROW_IF_NOT_OK(status) {auto localStatus = (status); if (localStatus) throw E_FAIL;}
+#define RETURN_HR_IF_NOT_OK(status) {auto localStatus = (status); if (localStatus) return E_FAIL;}
+
+template <typename T>
+using BaseType =
+std::remove_cv_t<
+    std::remove_reference_t<
+    std::remove_pointer_t<
+    std::remove_all_extents_t<T>
+    >
+    >
+>;
+
+template<typename T>
+using deleting_unique_ptr = std::unique_ptr<T, std::function<void(T*)>>;
+
+template <typename C, typename T = BaseType<decltype(*std::declval<C>().data())>>
+T GetElementCount(C const& range)
+{
+    return std::accumulate(range.begin(), range.end(), static_cast<T>(1), std::multiplies<T>());
+};
+
+// Create an ORT Session from a given model file path
+Ort::Session CreateSession(const wchar_t* model_file_path)
+{
+    OrtApi const& ortApi = Ort::GetApi();
+    const OrtDmlApi* ortDmlApi;
+    ortApi.GetExecutionProviderApi("DML", ORT_API_VERSION, reinterpret_cast<const void**>(&ortDmlApi));
+    Ort::Env ortEnvironment(ORT_LOGGING_LEVEL_WARNING, "DirectML_Direct3D_TensorAllocation_Test");
+    Ort::SessionOptions sessionOptions;
+    sessionOptions.SetExecutionMode(ExecutionMode::ORT_SEQUENTIAL);
+    sessionOptions.DisableMemPattern();
+    sessionOptions.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
+    ortApi.AddFreeDimensionOverrideByName(sessionOptions, "batch_size", 1);
+    OrtSessionOptionsAppendExecutionProvider_DML(sessionOptions, 0);
+
+    return Ort::Session(ortEnvironment, model_file_path, sessionOptions);
+}
+
+
+// Run the buffer through a preprocessing model that will shrink the
+// image from 512 x 512 x 4 to 224 x 224 x 3
+Ort::Value Preprocess(Ort::Session& session,
```

Comment:
[nit] Awkward to read with strange indent. Any of these will work:

a.
```
Ort::Value Preprocess(Ort::Session& session, ComPtr<ID3D12Resource> currentBuffer)
```

b.
```
Ort::Value Preprocess(
    Ort::Session& session,
    ComPtr<ID3D12Resource> currentBuffer
)
```

c.
```
Ort::Value Preprocess(Ort::Session& session,
                      ComPtr<ID3D12Resource> currentBuffer)
```

Appears (b) is the most self-consistent with surrounding code.